### PR TITLE
Iss 346 unet

### PR DESCRIPTION
**Pull request by a-arefina**
UNet as EcoderDecoder subclass
+ EncoderDecoder improvements

### CODE REVIEW

**Comment by roman-kh** [comment]
Let's move ` config['head/targets'] = self.get_from_attr('targets')` from `build_config` to `TFModel`

File: batchflow/models/tf/unet.py [hunk 1]
```diff
@@ -1,55 +1,59 @@
...
+        config['initial_block'] = dict(layout='cna cna', kernel_size=3, filters=64)
+
+        config['body/encoder/num_stages'] = 4
+        config['body/encoder/blocks'] = dict(layout='cna cna', kernel_size=3, filters=None)
+        config['body/embedding'] = None  # identity
+        config['body/decoder/num_stages'] = None  # defaults to body/encoder/num_stages
+        config['body/decoder/factor'] = None  # defaults to [2] * num_stages
+        config['body/decoder/blocks'] = dict(layout='cna cna', kernel_size=3, filters=None)
+
+        config['head'] = dict(layout='cna cna', kernel_size=[3, 3], strides=1, filters=64)
```
Comment (roman-kh) [inline batchflow/models/tf/unet.py]: Why not just 3?

Comment (a-arefina) [reply batchflow/models/tf/unet.py]: Legacy. I'll fix it

**Comment by a-arefina** [comment]
> Let's move ` config['head/targets'] = self.get_from_attr('targets')` from `build_config` to `TFModel`

'head/targets' is not needed in all `TFModel`s, only in `EncoderDecoder` and its children

**Comment by roman-kh** [comment]
> > Let's move ` config['head/targets'] = self.get_from_attr('targets')` from `build_config` to `TFModel`
> 'head/targets' is not needed in all `TFModel`s, only in `EncoderDecoder` and its children
That's right. But some models might use it. So having the same method signature would be great.

File: batchflow/models/tf/encoder_decoder.py [hunk 2]
```diff
@@ -66,6 +66,10 @@ class EncoderDecoder(TFModel):
                 other args : dict
                     Parameters for the base block.
 
+    head : dict
+        parameters for the head layers, usually :func:`.conv_block` parameters
+        an extra 1x1 convolutional layer that makes outputs shape match targets will be added implicitly if needed
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/encoder_decoder.py]:
```suggestion
        Note that an extra 1x1 convolution may be applied in order to make predictions compatible with the shape of targets
```

File: batchflow/models/tf/encoder_decoder.py [hunk 3]
```diff
@@ -146,11 +152,12 @@ def body(cls, inputs, name='body', **kwargs):
...
-            embeddings = embeddings if isinstance(embeddings, (tuple, list)) else [embeddings]
+            if embeddings is not None:
+                embeddings = embeddings if isinstance(embeddings, (tuple, list)) else [embeddings]
 
-            for i, embedding in enumerate(embeddings):
-                embedding_args = {**kwargs, **embedding}
-                x = cls.embedding(x, name='embedding'+str(i), **embedding_args)
+                for i, embedding in enumerate(embeddings):
+                    embedding_args = {**kwargs, **embedding}
+                    x = cls.embedding(x, name='embedding'+str(i), **embedding_args)
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/encoder_decoder.py]:
```suggestion
                    x = cls.embedding(x, name='embedding-'+str(i), **embedding_args)
```

File: batchflow/models/tf/encoder_decoder.py [hunk 4]
```diff
@@ -163,10 +170,16 @@ def body(cls, inputs, name='body', **kwargs):
     def head(cls, inputs, targets, name='head', **kwargs):
         """ Linear convolutions. """
         kwargs = cls.fill_params('head', **kwargs)
+
         with tf.variable_scope(name):
             x = cls.crop(inputs, targets, kwargs['data_format'])
+            x = super().head(x, name, **kwargs)
+
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/encoder_decoder.py]: You don't need extra scope named `name` inside already named `name` scope

Comment (a-arefina) [reply batchflow/models/tf/encoder_decoder.py]: Couldn't find better name, so leaving this as is

File: batchflow/models/tf/encoder_decoder.py [hunk 5]
```diff
@@ -20,16 +20,16 @@ class EncoderDecoder(TFModel):
         encoder : dict
             base : TFModel
                 Model implementing ``make_encoder`` method which returns tensors
-                with encoded representation of the inputs.
+                with encoded representation of the inputs. Optional
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/encoder_decoder.py]: move `optional` keyword to the first line of parameter description, e.g. `base : TFModel, optional`, just like [numpy style](https://numpydoc.readthedocs.io/en/latest/format.html) advises

File: batchflow/models/tf/encoder_decoder.py [hunk 6]
```diff
@@ -121,15 +130,11 @@ def default_config(cls):
...
-        config['body/decoder/blocks'] = dict(base=cls.block, combine_op='softsum')
-        config['head'] = dict(layout='c', kernel_size=1)
-        return config
+        config['body/decoder/blocks'] = dict(base=cls.block, combine_op='concat')
 
-    def build_config(self, names=None):
-        config = super().build_config(names)
-        config['head/targets'] = self.get_from_attr('targets')
-        return config
+        config['head/layout'] = None
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/encoder_decoder.py]: is this line necessary?

Comment (a-arefina) [reply batchflow/models/tf/encoder_decoder.py]: This is to explicitly show what the default parameter is

**Comment by roman-kh** [comment]
This closes #346

**Pull request merged**